fog/ledger/enclave/impl/src/lib.rs

```diff
@@ -93,30 +136,92 @@ impl LedgerEnclave for SgxLedgerEnclave {
+            .map_err(|e| {
+                log::error!(self.logger, "Could not decode user request: {}", e);
+                Error::ProstDecode
+            })?;
+
+        let mut resp = fog_types::ledger::CheckKeyImagesResponse {
+            num_blocks: untrusted_keyimagequery_response.highest_processed_block_count,
+            results: Default::default(),
+            global_txo_count: untrusted_keyimagequery_response.last_known_block_cumulative_count, /* ledger.num_txos */
+        };
 
-        // Try and deserialize.
-        let enclave_request: CheckKeyImagesRequest = mc_util_serial::decode(&request_bytes)?;
-        let mut key_images = Vec::new();
-        for query in enclave_request.queries {
-            key_images.push(query.key_image);
+        // Do the scope lock of keyimagetore
+        {
+            let mut lk = self.keyimagestore.lock()?;
+            let store = lk.as_mut().ok_or(Error::EnclaveNotInitialized)?;
+
+            resp.results = req
+                .queries
+                .iter() // Attempt and deserialize the untrusted portion of this request.
+                .map(|key| store.find_record(&key.key_image))
+                .collect();
         }
 
-        let context = KeyImageContext { key_images };
+        let response_plaintext_bytes = mc_util_serial::encode(&resp);
+
+        let response = self
+            .ake
+            .client_encrypt(&channel_id, &[], &response_plaintext_bytes)?;
 
-        Ok(context)
+        Ok(response.data)
     }
 
-    fn check_key_images_data(
-        &self,
-        response: CheckKeyImagesResponse,
-        client: ClientSession,
-    ) -> Result<EnclaveMessage<ClientSession>> {
-        // Serialize this for the client.
-        let response_bytes = mc_util_serial::encode(&response);
+    // Add a key image data to the oram sing the key image
+    fn add_key_image_data(&self, records: Vec<KeyImageData>) -> Result<()> {
+        let mut lk = self.keyimagestore.lock()?;
+        let store = lk.as_mut().ok_or(Error::EnclaveNotInitialized)?;
+        // add test KeyImageData record to ledger oram
+        for rec in records {
+            store.add_record(&rec.key_image, rec.block_index, rec.timestamp)?;
+        }
 
-        // Encrypt for the client.
-        Ok(self.ake.client_encrypt(&client, &[], &response_bytes)?)
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use fog_ledger_enclave_api::{messages::KeyImageData, AddRecordsError};
+    use mc_common::logger::create_root_logger;
+    use mc_transaction_core::ring_signature::KeyImage;
+    // Test that we were able to add key image record to the oram
+    #[test]
+    fn test_add_record() {
+        let desired_capacity: u64 = 1024 * 1024;
+        let logger = create_root_logger();
+        // create a new keyimagestore
+        let mut keyimagestore = key_image_store::KeyImageStore::<
+            mc_oblivious_traits::HeapORAMStorageCreator,
```

Comment:
nit: it's a little better to put `use mc_oblivious_traits::HeapORAMStorageCreator` at like line 199, so that this code is shorter and easier to read